src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+                if (foundIdx == -1)
+                {
+                    if (IsLastSpan)
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.EndOfStringNotFound);
+                    else
+                        return false;
+                }
+                if (foundIdx == 0)
+                    break;
+                for (int j = i + foundIdx - 1; j >= i; j--)
+                {
+                    if (_buffer[j] != JsonConstants.ReverseSolidus)
+                    {
+                        if ((counter & 1) == 0)
+                        {
+                            i += foundIdx;
+                            goto FoundEndOfString;
+                        }
+                        break;
+                    }
+                    else
+                        counter++;
+                }
+                i += foundIdx + 1;
+            }
+
+        FoundEndOfString:
+            int startIndex = _consumed + 1;
+            ReadOnlySpan<byte> localCopy = _buffer.Slice(startIndex, i - startIndex);
+
+            if (localCopy.IndexOfAnyControlOrEscape() != -1)
+            {
+                _lineBytePosition++;
+                ValidateEscapingAndHex(localCopy);
+                goto Done;
+            }
+
+            _lineBytePosition = i;
+
+        Done:
+            _lineBytePosition++;
+            ValueSpan = localCopy;
+            _tokenType = JsonTokenType.String;
+            _consumed = i + 1;
+            return true;
+        }
+
+        // https://tools.ietf.org/html/rfc8259#section-7
+        private void ValidateEscapingAndHex(ReadOnlySpan<byte> data)
+        {
+            // TODO: Optimize validating string contents
+            bool nextCharEscaped = false;
+            for (int i = 0; i < data.Length; i++)
+            {
+                byte currentByte = data[i];
+                if (currentByte == JsonConstants.ReverseSolidus)
+                    nextCharEscaped = !nextCharEscaped;
+                else if (nextCharEscaped)
+                {
+                    int index = JsonConstants.EscapableChars.IndexOf(currentByte);
+                    if (index == -1)
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.InvalidCharacterWithinString, currentByte);
+
+                    if (currentByte == 'n')
+                    {
+                        // Escaped new line character
+                        _lineBytePosition = -1; // Should be 0, but we increment _lineBytePosition below already
+                        _lineNumber++;
+                    }
+                    else if (currentByte == 'u')
+                    {
+                        // Expecting 4 hex digits to follow the escaped 'u'
+                        _lineBytePosition++;
+                        int startIndex = i + 1;
+                        for (int j = startIndex; j < data.Length; j++)
```

Comment:
If `data` does not include the terminating quote then I think that this will miss "\u012" (a \u near the end that doesn't have 4 trailing characters), but they get caught at any other place.

Reply:
Fixed in the new impl and added various tests.